bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java

```diff
@@ -0,0 +1,395 @@
+            authentication = null; // it should be null already, but for sake of consistency...
+            return false;
+        } else if (authentication != null) {
+            // We are already authenticated, nothing to do.
+            // If you want to force re-authentication, set this.authentication = null.
+            return true;
+        }
+
+        try {
+            final SOAPMessage initialAuthPayload = newMessage("Login", newSOAPElement("Action", "request"),
+                    newSOAPElement("Username", USER), newSOAPElement("LoginPassword"), newSOAPElement("Captcha"));
+
+            final Document challengeDocument = post("Login", initialAuthPayload);
+            final String challenge = getElementByTagName(challengeDocument, "Challenge");
+            final String cookie = getElementByTagName(challengeDocument, "Cookie");
+            final String publicKey = getElementByTagName(challengeDocument, "PublicKey");
+            final String privateKey = calculateHMAC(challenge, publicKey + pin);
+            final String loginPassword = calculateHMAC(challenge, privateKey);
+
+            authentication = new Authentication(privateKey, cookie);
+            final SOAPMessage payload = newMessage("Login", newSOAPElement("Action", "login"),
+                    newSOAPElement("Username", USER), newSOAPElement("LoginPassword", loginPassword),
+                    newSOAPElement("Captcha"));
+            final Document loginDocument = post("Login", payload, authentication);
+            final String loginStatus = getElementByTagName(loginDocument, "LoginResult").toLowerCase();
+            if ("success".equals(loginStatus)) {
+                updateStatus(ThingStatus.ONLINE);
+                invalidConfiguration = false;
+                return true;
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Failed to authenticate.");
+                invalidConfiguration = true;
+                authentication = null;
+                return false;
+            }
+        } catch (CommunicationException e) {
+            authentication = null;
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Failed to authenticate.");
+            return false;
+        } catch (SOAPException e) {
+            authentication = null;
+            logger.error("Unexpected internal error.");
+            updateStatus(ThingStatus.OFFLINE);
+            return false;
+        }
+    }
+
+    private Document post(final String action, final SOAPMessage message) throws CommunicationException {
+        return post(action, message, null);
+    }
+
+    private Document post(final String action, final SOAPMessage message, final Authentication authentication)
+            throws CommunicationException {
+        Request request = httpClient.newRequest(url);
+        request.timeout(5, TimeUnit.SECONDS);
```

Comment:
Is 5 sec. sufficient under all circumstances? Most bindings use a timeout of 10 sec.